lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js

```diff
@@ -76,7 +77,7 @@ describe('LMSGrader', () => {
     );
   });
 
-  it('set the selected student count to "Student 2 of 2" when the index changers to 1', () => {
+  it('set the selected student count to "Student 2 of 2" when the index changers to 1', async () => {
```

Comment:
@lyzadanger 

Sinon drives me nuts sometimes!

Since this technically calls a promise (look at line 86), we seem to need to make the function `async` even though I don't need to `await` on that promise or even care about it returning because the stuff I'm testing here is not dependent on that to resolve, but if I don't make the test `async`, the spy `fakeRpcCall` seems to not work as expected in future tests. Specifically the  `restHistory` does not reset that spy's history and I observed that the call count was growing in spite of calling `restHistory`

I think what was happening is the first test didn't finish until after the second test started? Anyway so you'll notice if you remove `async` from this test, the next test fails because the callcount is 3 instead of 1 and `calledOnceWithExactly` is never true.

Reply:
> I think what was happening is the first test didn't finish until after the second test started? 

That seems possible. My immediate hypothesis about what is going on is that it is to do with the fact that async functions always return a Promise:

Making this function async will cause it to return a Promise, even though there is no explicit return in the function. The promise will "immediately" resolve to `undefined`. Since the result is a Promise, mocha will wait for it to resolve before continuing to the next test. It does this waiting by queuing a microtask to run once all the promises resolve. This microtask will run _after_ any other tasks in the queue.

If the function returns synchronously on the other hand, the next test will run immediately without the microtask queue being flushed.

If this test results in some microtasks being scheduled but not yet executed by the time the test returns, the tasks could end up running at some later point and manipulating shared state in those other tasks.

An alternative fix is to change the usage of sinon so that each test gets a fresh stub, rather than sharing the same stub instance between tests and using `resetHistory`. This change would ensure that any code which happens to run after the test returns won't end up affecting future tests.

Reply:
Thanks Rob. That makes sense, I think as best practice, we always re-create the stubs (if they change) but always the spies each test.